Data command seperation

s-axford opened the pull request:
All shared code is now in Message, DataMessages now have a request section, CommandMessage is implemented, and all messages have a flag field. (May move these into a structure like power keys if we plan to keep adding message types)

s-axford committed:
Separated shared code into Message, and implemented CommandMessage

s-axford committed:
Added flag to indicate message type

liamhebert commented on internal-messaging/message/cc/DataMessage.cc:
```diff
@@ -110,75 +59,15 @@ DataMessage::DataMessage(char* flat)
-        memset(hex_string, 0, 32);
-        i++;
-    }
+    i = BuildContents(flat, i);
```
Hmm... It seems here that the behavior of `flatten` slightly is different between `DataMessage` and `CommandMessage` where `BuildContents` is called in the constructor of `DataMessage` compared to in `CommandMessage` where it is called in `flatten()`. Is this desired? It might be ideal to move `BuildContents` to either both be in the constructor or both in the `flatten()`

s-axford replied:
BuildContents should be used in the constructors, not Flatten. However I don't see a BuildContents in the Flatten method of CommandMessage. am I missing something? 

BuildContents is called on line 38 of CommandMessage.cc, which is within its constructor

liamhebert replied:
Ah my bad, I must have read this too quickly. You are correct here sorry

liamhebert reviewed (approved):
This is interesting that we are shifting from having generic `Message`'s to having message types. I think it would be worth declaring all or most of the methods in `Message` as `virtual` to force usage of these new types. As well, I believe it would be worth it to implement a sort of [changelog](https://keepachangelog.com/en/1.0.0/) with [Semantic Versioning](https://semver.org/) since this library is used by other applications. This is getting a bit more important due to API breaking changes like this by shifting to concrete message types and will allow other teams to know when to update their methods. Either way, this is awesome :)

s-axford commented:
> This is interesting that we are shifting from having generic `Message`'s to having message types. I think it would be worth declaring all or most of the methods in `Message` as `virtual` to force usage of these new types. As well, I believe it would be worth it to implement a sort of [changelog](https://keepachangelog.com/en/1.0.0/) with [Semantic Reasoning](https://semver.org/) since this library is used by other applications. This is getting a bit more important due to API breaking changes like this by shifting to concrete message types. Either way, this is awesome :)

My reasoning for leaving most of the methods without a virtual keyword is because I don't want people to override those implementations. However Flatten is now a pure virtual function which will force people to use a child class (as you cannot create an object of a pure virtual class). A changelog would be a great idea. I will mention that next scrum period.

liamhebert commented:
> However Message's Flatten function is now a pure virtual function which will force people to use a child class (as you cannot create an object of a pure virtual class). 

Missed that, good stuff! My apologies for my missing these points, I'm still getting familiar with C++ compared to being in a Java world.

s-axford commented:
Also, as of this PR the current message types look like this:
DataMessage: 
sender|recipient|time_created|flag|request-request-request|key~value|key~value|\0
flag = 0

CommandMessage:
sender|recipient|time_created|flag|commandkey~value|commandkey~value|\0
flag = 1

(Message is now purely virtual) Currently ALL existing code will utilizes DataMessage, CommandMessage functionality will be added soon

I believe an official document is coming soon that will record this, and any changes to the message structure.

s-axford committed:
Added comments outlining format of each message type

AndrewWay commented on internal-messaging/message/cc/CommandMessage.cc:
```diff
@@ -0,0 +1,55 @@
+CommandMessage::CommandMessage()
+:Message()
+{
+    flag = 1;
```
Variables such as "flag" should **always** be stored in constant variables, and defined only once.

s-axford replied:
Im not sure what a better way to initialize flag would be. flag is a protected member of Message, not specific to DataMessage. So it cannot be initialized in the DataMessage.h header. I  could make it a parameter of Messages constructor which I am calling, but I am not sure that is better.

Can you suggest a implementation?

AndrewWay commented on internal-messaging/message/cc/DataMessage.cc:
```diff
@@ -3,96 +3,45 @@
 #include "DataMessage.h"
 #include <stdexcept>
 
-bool number_value(const char* string) {
```
It's not immediately obvious what this function does. Can you add some comments?

AndrewWay commented on internal-messaging/message/cc/DataMessage.cc:
```diff
@@ -3,96 +3,45 @@
 :Message()
 {
-    // NULL
+    flag = 0;
```
Store in constant variable, defined only once.

AndrewWay commented on internal-messaging/message/cc/DataMessage.cc:
```diff
@@ -3,96 +3,45 @@
+   flag = 0;
 }
 
 DataMessage::DataMessage(char* flat)
```
Constructor is very long. Anyway to shorten this and allow for reusability in functions

AndrewWay replied:
By shorten I mean break up into functions

AndrewWay commented on internal-messaging/message/cc/DataMessage.cc:
```diff
@@ -110,75 +59,15 @@ DataMessage::DataMessage(char* flat)
+    i = BuildContents(flat, i);
 }
 
 void DataMessage::Flatten(char* msg) {
```
Could use some comments.

AndrewWay commented on internal-messaging/message/cc/Message.cc:
```diff
@@ -3,6 +3,23 @@
 #include "Message.h"
 #include <stdexcept>
 
+bool number_value(const char* string) {
```
Comments

AndrewWay reviewed (changes requested):
I see the flags for data and commands are 0 and 1. Should probably make them more distinguishable. All it takes is a single bit flip and a data message becomes a command... freaky stuff might ensue. 

Other than that it looks good.

s-axford committed:
Added comment on number_value, and changed flag values

s-axford committed:
Fixing flag parse to suit new flags

s-axford commented:
I have changed the flag values from 0 and 1 to 100 and 200. Since the data is converted to hex this will be 64 and C8 and will not switch with a bit swap. I also added comments on the number_value function

s-axford committed:
Fixed multiple declarations of flag

s-axford committed:
Did not mean to change LorisMessager.cc, simplified base constructors, added check for correct flag in constructor with flattened arrays
